jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java

```diff
@@ -58,45 +63,86 @@ static JavaLayerConfigurations getForProject(
     List<Path> extraFiles = new ArrayList<>();
+    final boolean isWebapp = GradleProjectProperties.getWarTask(project) != null;
 
-    // Adds each file in each classes output directory to the classes files list.
-    FileCollection classesOutputDirectories = mainSourceSet.getOutput().getClassesDirs();
-    gradleJibLogger.info("Adding corresponding output directories of source sets to image");
-    for (File classesOutputDirectory : classesOutputDirectories) {
-      if (Files.notExists(classesOutputDirectory.toPath())) {
-        gradleJibLogger.info("\t'" + classesOutputDirectory + "' (not found, skipped)");
-        continue;
+    if (isWebapp) {
+      if (appRoot.isEmpty()) {
+        appRoot = ContainerConfiguration.DEFAULT_WEBAPP_ROOT;
       }
-      gradleJibLogger.info("\t'" + classesOutputDirectory + "'");
-      try (Stream<Path> classFileStream = Files.list(classesOutputDirectory.toPath())) {
+      Path explodedWarPath = GradleProjectProperties.getExplodedWarDirectory(project).toPath();
+      Path classesOutputDirectory =
+          explodedWarPath.resolve(JavaLayerConfigurations.WEB_INF_CLASSES);
+      try (Stream<Path> classFileStream = Files.list(classesOutputDirectory)) {
         classFileStream.forEach(classesFiles::add);
       }
-    }
-    if (classesFiles.isEmpty()) {
-      gradleJibLogger.warn("No classes files were found - did you compile your project?");
-    }
 
-    // Adds each file in the resources output directory to the resources files list.
-    Path resourcesOutputDirectory = mainSourceSet.getOutput().getResourcesDir().toPath();
-    if (Files.exists(resourcesOutputDirectory)) {
-      try (Stream<Path> resourceFileStream = Files.list(resourcesOutputDirectory)) {
-        resourceFileStream.forEach(resourcesFiles::add);
+      Path libOutputDirectory = explodedWarPath.resolve(JavaLayerConfigurations.WEB_INF_LIB);
+      try (Stream<Path> dependencyFileStream = Files.list(libOutputDirectory)) {
+        dependencyFileStream.forEach(
+            path -> {
+              if (path.toString().contains(SNAPSHOT)) {
+                snapshotDependenciesFiles.add(path);
+              } else {
+                dependenciesFiles.add(path);
+              }
+            });
+      }
+      // All files except classes and libs in resources
+      try (Stream<Path> fileStream = Files.list(explodedWarPath)) {
+        fileStream
+            .filter(path -> !path.endsWith(JavaLayerConfigurations.WEB_INF))
+            .forEach(resourcesFiles::add);
+      }
+      // Some files in WEB-INF need to be in resources directory (e.g. web.xml, ...)
+      Path webinfOutputDirectory = explodedWarPath.resolve(JavaLayerConfigurations.WEB_INF);
+      try (Stream<Path> fileStream = Files.list(webinfOutputDirectory)) {
+        fileStream
+            .filter(path -> !path.endsWith("classes") && !path.endsWith("lib"))
+            .forEach(resourcesFiles::add);
+      }
+    } else {
+      if (appRoot.isEmpty()) {
```

Comment:
Can this happen? I thought `appRoot` is `DEFAULT_APP_ROOT` when not specified. Can you verify?

Reply:
There is 2 default values : DEFAULT_WEBAPP_ROOT and DEFAULT_APP_ROOT.

So, I changed this code 

```
ContainerParameters {
    private String appRoot = DEFAULT_APP_ROOT;
}
```
with that : 

```
ContainerParameters {
    private String appRoot = "";
}
```

Reply:
Got it. We'll worry about this (if we have to) later.